### PR TITLE
RFC 5425

### PR DESCRIPTION
**Pull request by leodido**

### CODE REVIEW

**Comment by leodido** [comment]
I think this code is in a good state now. But waiting for my friend @goller to review it.

File: rfc5424/syslog_message.go [hunk 1]
```diff
@@ -88,15 +88,27 @@ func (sm *SyslogMessage) Valid() bool {
...
+//
+// It enforces a correct priority value (range [0, 191]), since it can not by null by RFC.
+func (sm *SyslogMessage) SetPriority(value uint8) *SyslogMessage {
+	if value >= 0 && value <= 191 {
+		sm.setPriority(value)
+
+		return sm
+	}
+
+	panic("out of range priority value")
```
Comment (goller) [inline rfc5424/syslog_message.go]: I don't think this should panic.

I think you would mark the message as invalid or return a nil or something.

File: rfc5425/parser.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	"github.com/influxdata/go-syslog/rfc5424"
+)
+
+// Parser represents a parser
+type Parser struct {
+	s          Scanner
+	p          rfc5424.Parser
+	bestEffort bool
+	msglen     int64
+	buf        struct {
```
Comment (goller) [inline rfc5425/parser.go]: I think you should simplify this by flattening without a struct.

File: rfc5425/parser.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		tok Token // last read token
+		num int   // size (max = 1)
+	}
+}
+
+// NewParser returns a new instance of Parser
+func NewParser(r io.Reader, bestEffort bool) *Parser {
```
Comment (goller) [inline rfc5425/parser.go]: I really don't like bool parameters... looks nice from the library but not the library user

```go
NewParser(r, true)
```
is hard to know what is going on.

I think you should use an opts pattern similar to what I did with telegraf.

File: rfc5425/parser.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+// Result represent the resulting syslog message and (eventually) error occured during parsing
+type Result struct {
+	Message      *rfc5424.SyslogMessage
+	MessageError error
+	Error        error
+}
+
+// ResultHook is a function the user can use to specify what to do with every `Result` instance
+type ResultHook func(result *Result)
```
Comment (goller) [inline rfc5425/parser.go]: I think this should be `ResultHandler`

**Review by goller — COMMENTED** [review]
Looks good... I have some suggestions for minor changes

File: rfc5424/syslog_message.go (hunk 1 above)
Comment (leodido) [reply rfc5424/syslog_message.go]: Chosen to ignore out of range values simply not setting them.

This setters always return a `SyslogMessage` to have a fluent style.

**Comment by leodido** [comment]
I should have done everything you mentioned @goller :)

Please notice that the builder here introduced is simply an initial version.
The idea is to create a builder capable to build `SyslogMessage` valid instances, capable to serialize each of them in a string as per RFC5424. Doing this we can make fields private.
Anyway to do such a builder I intend to exploit the ragel pieces (but I'll need a modularization of the ragel machine). This way only valid (syntactically and semantically in some cases) inputs will be accepted.
It is possible to achieve this, I already done a preliminary try about this. But it is out of the scope of this PR.

**Pull request merged**